molecule/src/structure_checker2.cpp

```diff
@@ -0,0 +1,651 @@
+    StructureChecker2::CHECK_CHARGE_TXT,    StructureChecker2::CHECK_SALT_TXT,         StructureChecker2::CHECK_AMBIGUOUS_H_TXT,
+    StructureChecker2::CHECK_COORD_TXT,     StructureChecker2::CHECK_V3000_TXT,        StructureChecker2::CHECK_ALL_TXT};
+
+static CheckParams check_params_from_string(const char* params)
+{
+    CheckParams r;
+    size_t len = sizeof(checkTypeName) / sizeof(*checkTypeName);
+    if (params)
+    {
+        std::smatch sm1;
+        std::unordered_set<std::string> words;
+        std::string s(params);
+        std::regex rx1(R"(\b(\w+)\b)", std::regex_constants::icase);
+        while (std::regex_search(s, sm1, rx1))
+        {
+            words.insert(sm1[1]);
+            s = sm1.suffix();
+        }
+        r.check_flags = StructureChecker2::CHECK_NONE;
+        for (int i = 0; i < len; i++)
+        {
+            if (words.find(std::string(checkTypeName[i])) != words.end())
+            {
+                r.check_flags |= 1 << i;
+            }
+        }
+        r.check_flags = !r.check_flags || r.check_flags & 1 << (len - 1) ? StructureChecker2::CHECK_ALL
+                                                                         : (r.check_flags == 1 ? StructureChecker2::CHECK_NONE : r.check_flags >> 1);
+
+        std::smatch sm2;
+        s = params;
+        std::regex rx2(R"(\b(atoms|bonds)\b((?:\W+\b\d+\b\W*?)+))", std::regex_constants::icase);
+        std::regex rx3(R"(\b(\d+)\b)");
+        std::smatch sm3;
+        while (std::regex_search(s, sm2, rx2))
+        {
+            std::vector<int>& vec = std::tolower(sm2[1].str()[0]) == 'a' ? r.selected_atoms : r.selected_bonds;
+            std::string a = sm2[2];
+            while (std::regex_search(a, sm3, rx3))
+            {
+                vec.push_back(atoi(sm3[1].str().c_str()));
+                a = sm3.suffix();
+            }
+            s = sm2.suffix();
+        }
+    }
+    return r;
+}
+static void checkMolecule(const IndigoObject& bmol, int check_types, const std::vector<int>& selected_atoms, const std::vector<int>& selected_bonds,
+                          StructureChecker2::CheckResult& result);
+static void checkMolecule(BaseMolecule& item, int check_types, const std::vector<int>& selected_atoms, const std::vector<int>& selected_bonds,
+                          StructureChecker2::CheckResult& result);
+
+static const std::string& message_text(StructureChecker2::CheckMessageCode code);
+
+StructureChecker2::StructureChecker2()
+{
+}
+
+StructureChecker2::CheckResult StructureChecker2::check(const char* item, const char* check_flags, const char* load_params)
+{
+    int it = indigoLoadStructureFromString(item, load_params);
+    if (it < 0)
+    {
+        it = indigoLoadStructureFromString(item, (std::string(load_params ? load_params : "") + " query").c_str()); //##!!!PATCH
+    }
+    auto r = check(it, check_flags);
+    indigoFree(it);
+    return r;
```

Comment:
Same here as in below comment - we should not use Indigo API in core libraries. We need to load molecule from string in API and pass `BaseMolecule` here instead of string.